pkg/telnetlib/connection.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telnetlib
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type connOpts struct {
+	conn        net.Conn
+	fsm         *telnetFSM
+	cmdHandler  CmdHandlerFunc
+	dataHandler DataHandlerFunc
+	serverOpts  map[byte]bool
+	clientOpts  map[byte]bool
+	optCallback func(byte, byte)
+}
+
+type TelnetConn struct {
+	conn              net.Conn
+	readCh            chan []byte
+	writeCh           chan []byte
+	unackedServerOpts map[byte]bool
+	unackedClientOpts map[byte]bool
+	//server            *TelnetServer
+	serverOpts         map[byte]bool
+	clientOpts         map[byte]bool
+	dataRW             io.ReadWriter
+	cmdBuffer          bytes.Buffer
+	fsm                *telnetFSM
+	fsmInputCh         chan byte
+	handlerWriter      io.Writer
+	cmdHandler         CmdHandlerFunc
+	dataHandler        DataHandlerFunc
+	dataHandlerCloseCh chan chan struct{}
+	dataWrittenCh      chan bool
+	optionCallback     func(byte, byte)
+	connReadDoneCh     chan chan struct{}
+	connWriteDoneCh    chan chan struct{}
+	negotiationDone    chan struct{}
+	closedMutex        sync.Mutex
+	closed             bool
+}
+
+// Safely read/write concurrently to the data Buffer
+// databuffer is written to by the FSM and it is read from by the dataHandler
+type dataReadWriter struct {
+	dataBuffer bytes.Buffer
+	dataMux    *sync.Mutex
+}
+
+func (drw *dataReadWriter) Read(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Read(p)
+}
+
+func (drw *dataReadWriter) Write(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Write(p)
+}
+
+type connectionWriter struct {
+	ch chan []byte
+}
+
+func (cw *connectionWriter) Write(b []byte) (int, error) {
+	cw.ch <- b
```

Comment:
it should get closed when the telnet connection closes. I changed the TelnetConnection.handlerWriter to be an io.WriteCloser instead of io.Writer. When the TelnetConn.Close() is called, it closes this writer which in turns closes the channel and set it to nil. In the code above, I will always check that cw.ch is not nil before writing to it.